ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c

```diff
@@ -0,0 +1,94 @@
+/* -*- Mode: C; c-basic-offset:4 ; -*- */
+/*
+ * Copyright (c) 2014      Argonne National Laboratory.
+ * Copyright (c) 2019      Research Organization for Information Science
+ *                         and Technology (RIST).  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "cdesc.h"
+
+int cdesc_create_datatype(CFI_cdesc_t *cdesc, int oldcount, MPI_Datatype oldtype, MPI_Datatype *newtype)
+{
+    const int MAX_RANK = 15; /* Fortran 2008 specifies a maximum rank of 15 */
+    MPI_Datatype types[MAX_RANK + 1]; /* Use a fixed size array to avoid malloc. + 1 for oldtype */
+    int mpi_errno = MPI_SUCCESS;
+    int accum_elems = 1;
+    int accum_sm = cdesc->elem_len;
+    int done = 0; /* Have we created a datatype for oldcount of oldtype? */
+    int last; /* Index of the last successfully created datatype in types[] */
+    int extent;
+    int i, j;
+
+#ifdef OPAL_ENABLE_DEBUG
+    {
+        int size;
+        assert(cdesc->rank <= MAX_RANK);
+        ompi_datatype_size(oldtype, &size);
+        /* When cdesc->elem_len != size, things suddenly become complicated. Generally, it is hard to create
+         * a composite datatype based on two datatypes. Currently we don't support it and doubt it is usefull.
+         */
+        assert(cdesc->elem_len == size);
+    }
+#endif
+
+    types[0] = oldtype;
+    i = 0;
+    done = 0;
+    while (i < cdesc->rank && !done) {
+        if (oldcount % accum_elems) {
+            /* oldcount should be a multiple of accum_elems, otherwise we might need an
+             * MPI indexed datatype to describle the irregular region, which is not supported yet.
+             */
+            mpi_errno = MPI_ERR_INTERN;
+            last = i; goto fn_exit;
+        }
+
+        extent = oldcount / accum_elems;
+        if (extent > cdesc->dim[i].extent) {
+            extent = cdesc->dim[i].extent;
+        } else {
+            /* Up to now, we have accumlated enough elements */
+            done = 1;
+        }
+
+        if (cdesc->dim[i].sm == accum_sm) {
+            mpi_errno = PMPI_Type_contiguous(extent, types[i], &types[i+1]);
+        } else {
+            mpi_errno = PMPI_Type_create_hvector(extent, 1, cdesc->dim[i].sm, types[i], &types[i+1]);
+        }
+        if (mpi_errno != MPI_SUCCESS) {
+            last = i; goto fn_exit;
+        }
+
+        mpi_errno = PMPI_Type_commit(&types[i+1]);
```

Comment:
Don't we only have to commit the *last* datatype?

I.e., isn't "commit" potentially an expensive operation?  If so, can't we use uncommitted datatypes to build up new datatypes, and therefore only have to commit the *last* datatype?

Reply:
@bosilca confirmed to me in an email that I was correct -- you only need to commit the final datatype.  Depending on the complexity of the array slice, this could be a nice optimization.

Reply:
yep, I will do that (and issue a PR to mpich since this is basically their code).

Reply:
Still need to move this `type_commit` outside of the loop and only commit the last/final datatype.